server/game/core/ability/PlayCardAction.ts

```diff
@@ -6,14 +6,22 @@ import { KeywordName, PhaseName, PlayType, Stage } from '../Constants';
 import { ICost } from '../cost/ICost';
 import { AbilityContext } from './AbilityContext';
 import PlayerAction from './PlayerAction';
+import { TriggerHandlingMode } from '../event/EventWindow.js';
+import { CostAdjuster } from '../cost/CostAdjuster';
 
 export type PlayCardContext = AbilityContext & { onPlayCardSource: any };
 
 export abstract class PlayCardAction extends PlayerAction {
-    protected playType: PlayType;
-
-    public constructor(card: Card, title: string, playType: PlayType, additionalCosts: ICost[] = [], targetResolver: IActionTargetResolver = null) {
-        super(card, PlayCardAction.getTitle(title, playType), additionalCosts.concat(CostLibrary.payPlayCardResourceCost(playType)), targetResolver);
+    public constructor(
+        card: Card,
+        title: string,
+        protected playType: PlayType,
+        triggerHandlingMode: TriggerHandlingMode = TriggerHandlingMode.ResolvesTriggers,
+        public costAdjuster: CostAdjuster = null,
```

Comment:
Needs readonly, so does `playType` actually. Also I'm realizing I don't love declaring properties in the ctor params like this, it's actually kind of hard to notice them imo and decreases overall readability. If you agree then would you mind creating these as explicit properties and just doing like `this.playType = playType` in the ctor? If you like it this way then feel free to keep it.